bumble/hid.py

```diff
@@ -0,0 +1,301 @@
+        FEATURE_REPORT      = 0x03
+
+    # Handshake parameters
+    class Handshake(enum.IntEnum):
+        SUCCESSFUL              = 0x00
+        NOT_READY               = 0x01
+        ERR_INVALID_REPORT_ID   = 0x02
+        ERR_UNSUPPORTED_REQUEST = 0x03
+        ERR_UNKNOWN             = 0x0E
+        ERR_FATAL               = 0x0F
+
+    #Message Type
+    class MessageType(enum.IntEnum):
+        HANDSHAKE           = 0x00
+        CONTROL             = 0x01
+        GET_REPORT          = 0x04
+        SET_REPORT          = 0x05
+        GET_PROTOCOL        = 0x06
+        SET_PROTOCOL        = 0x07
+        DATA                = 0x0A
+
+    # Protocol modes
+    class ProtocolMode(enum.IntEnum):
+        BOOT_PROTOCOL      = 0x00
+        REPORT_PROTOCOL    = 0x01
+
+    # Control Operations
+    class ControlCommand(enum.IntEnum):
+        SUSPEND              = 0x03
+        EXIT_SUSPEND         = 0x04
+        VIRTUAL_CABLE_UNPLUG = 0x05
+
+
+# HIDP messages
+@dataclass
+class GetReportMessage(Message):
+    report_type : int
+    report_id : int
+    buffer_size : int
+
+    def __bytes__(self) -> bytes:
+        if(self.report_type == Message.ReportType.OTHER_REPORT):
+            param = self.report_type
+        else:
+            param = 0x08 | self.report_type
+        header = ((Message.MessageType.GET_REPORT << 4) | param)
+        packet_bytes = bytearray()
+        packet_bytes.append(header)
+        packet_bytes.append(self.report_id)
+        packet_bytes.extend([(self.buffer_size & 0xff), ((self.buffer_size >> 8) & 0xff)])
+        return bytes(packet_bytes)
+@dataclass
+class SetReportMessage(Message):
+    report_type: int
+    data : bytes
+
+    def __bytes__(self) -> bytes:
+        header = ((Message.MessageType.SET_REPORT << 4) | self.report_type)
+        packet_bytes = bytearray()
+        packet_bytes.append(header)
+        packet_bytes.extend(self.data)
+        return bytes(packet_bytes)
+@dataclass
+class GetProtocolMessage(Message):
+
+    def __bytes__(self) -> bytes:
+        header = (Message.MessageType.GET_PROTOCOL << 4)
+        packet_bytes = bytearray()
+        packet_bytes.append(header)
+        return bytes(packet_bytes)
+@dataclass
+class SetProtocolMessage(Message):
+    protocol_mode: int
+
+    def __bytes__(self) -> bytes:
+        header = (Message.MessageType.SET_PROTOCOL << 4 | self.protocol_mode)
+        packet_bytes = bytearray()
+        packet_bytes.append(header)
+        packet_bytes.append(self.protocol_mode)
+        return bytes(packet_bytes)
+@dataclass
+class SendData(Message):
+
+    data : bytes
+
+    def __bytes__(self) -> bytes:
+        header = ((Message.MessageType.DATA << 4) | Message.ReportType.OUTPUT_REPORT)
+        packet_bytes = bytearray()
+        packet_bytes.append(header)
+        packet_bytes.extend(self.data)
+        return bytes(packet_bytes)
+# -----------------------------------------------------------------------------
```

Comment:
missing two blank lines above the separator

Reply:
Added